### DISCUSSION TITLE
Serverless Functions `type: "module"` got error

### DISCUSSION THREAD

**Discussion by Lmmmmmm-bb** [vercel/community#1225] [Help] [answered]
### Summary

I have a Telegram bot deploy by Vercel Serverless Functions.
Everything was fine when I deployed it before, but recently I found an error when I redeployed it.

1. I used Edge Functions to deploy in this [commit](https://github.com/Lmmmmmm-bb/m-record-bot/commit/8deaacfd9696039b2ddb080b8b017ed6efc7afb7). 
2. I found the error in the **Edge Functions** when I deployed it after committing, **I reverted the code**. [relate commit](https://github.com/Lmmmmmm-bb/m-record-bot/commit/4921686d484be12b93c895de7c2ad6e8600cb639)
3. After the second step, **the code should be unchanged compared to before. However, after redeploying the Telegram bot, the following error occurs.** [relate commit](https://github.com/Lmmmmmm-bb/m-record-bot/commit/6d3580c31de7e79332ec8fdc43978bb087084be3)
![image](https://user-images.githubusercontent.com/54026110/211188071-d114aea3-5821-4a6e-a270-f05aee1cdbe9.png)

Returned to normal after I delete the package.json **`type: "module"`** field.

I am curious why the bot did not get this error when it was deployed before.

This is repository https://github.com/Lmmmmmm-bb/m-record-bot. Any help is greatly appreciated!

**Comment by duiker101** [comment]
I think I have a related issue.

My app doesn't use `"type":"module"` and on current deploy (same code as before) now I get this.

![WuE0F8NUsz](https://user-images.githubusercontent.com/755683/211198072-a128c31a-a54d-4b6d-95bf-ea95be58239a.png)

Adding the `"type":"module"` still breaks with this error

![image](https://user-images.githubusercontent.com/755683/211198650-d6d3faf7-7f3a-4d3e-af73-5abbf4d2e181.png)

Dev environment seems to work ok. This is preventing me from deploying ATM so any solution would be appreciated.

**Comment by tmilar** [comment]
I started having the same issue since Friday, for any new deployment I do since that moment.
Looks like any Typescript API function stopped working, as if something broke in the build process in Vercel side.

Even a very basic function like this doesn't work for me:
```ts
// ./api/hello.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'

export default function (req: VercelRequest, res: VercelResponse) {
  const { name = 'World' } = req.query
  res.send(`Hello ${name}!`)
}
```

Results in response:
<img width="634" alt="image" src="https://user-images.githubusercontent.com/3988039/211202365-52b4dd27-9e7e-44dd-9a24-1a962cd5d4e0.png">

The logs show: 
<img width="1212" alt="image" src="https://user-images.githubusercontent.com/3988039/211202399-5c062437-c27d-4834-9516-49ae491d5867.png">

My tsconfig.json file:

```json
{
  "compilerOptions": {
    "target": "es6",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "strict": true,
    "pretty": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react"
  },
  "include": ["src"],
  "exclude": ["**/node_modules", "**/.*/", "dist", "public", "test", "api"]
}
```

My package.json:

```json
{
  "name": "my-project",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "start": "react-scripts start",
    "start:server": "cd api && npm start",
    "start:vercel": "vercel dev",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "lint": "tsc --noEmit && eslint . --ext \"ts,tsx\"",
    "lint:fix": "npm run lint -- --fix",
    "clean": "rm -rf dist"
  },
  "dependencies": {
    "@auth0/auth0-react": "1.1.0",
    "@vercel/node": "1.10.0",
    "auth0-js": "9.14.0",
    "axios": "^0.27.2",
    "connected-react-router": "6.8.0",
    "date-fns": "^2.28.0",
    "got": "11.8.2",
    "history": "^4.10.1",
    "i18next": "^21.6.3",
    "i18next-browser-languagedetector": "^6.1.2",
    "jsonwebtoken": "8.5.1",
    "jwks-rsa": "2.0.3",
    "lodash": "4.17.20",
    "query-string": "6.13.6",
    "react": "16.13.1",
    "react-csv": "2.0.3",
    "react-dom": "16.13.1",
    "react-i18next": "^11.15.1",
    "react-redux": "7.2.1",
    "react-router-dom": "5.2.0",
    "react-scripts": "3.4.3",
    "recharts": "2.0.9",
    "redux-logger": "3.0.6",
    "redux-saga": "1.1.3",
    "semantic-ui-css": "2.4.1",
    "semantic-ui-react": "1.3.1",
    "typesafe-actions": "5.1.0"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "5.11.4",
    "@testing-library/react": "11.0.4",
    "@testing-library/user-event": "12.1.6",
    "@types/auth0-js": "9.14.0",
    "@types/jest": "26.0.14",
    "@types/jsonwebtoken": "^8.5.6",
    "@types/lodash": "4.14.164",
    "@types/node": "14.11.2",
    "@types/react": "16.9.49",
    "@types/react-csv": "1.1.2",
    "@types/react-dom": "16.9.8",
    "@types/react-redux": "7.1.9",
    "@types/react-router-dom": "5.1.5",
    "@types/redux-logger": "3.0.8",
    "@types/segment-analytics": "^0.0.34",
    "@typescript-eslint/eslint-plugin": "4.3.0",
    "@typescript-eslint/parser": "4.3.0",
    "eslint": "6.8.0",
    "eslint-config-prettier": "6.12.0",
    "eslint-plugin-import": "2.22.1",
    "eslint-plugin-jsx-a11y": "6.3.1",
    "eslint-plugin-prettier": "3.1.4",
    "eslint-plugin-react": "7.21.2",
    "prettier": "2.1.2",
    "typescript": "4.0.3"
  },
}

```

**Comment by steelballs** [comment]
I had the same issue on Friday too. It seems to have started occurring after an update to vercel CLI which is used during the deployment.

This deployment worked as expected.

<img width="926" alt="Screenshot 2023-01-07 at 17 54 21" src="https://user-images.githubusercontent.com/6048892/211204069-2421ce04-5d2d-4f00-837d-eda9ec5b7efb.png">

This deployment broke.

<img width="927" alt="Screenshot 2023-01-07 at 17 55 09" src="https://user-images.githubusercontent.com/6048892/211204070-c3a4e69e-6f47-4587-ba92-5f446df10fd4.png">

**Comment by krystof-k** [comment]
Yes, the same for me, also happening since Friday. Same commit built now results in the following error:

```console
[POST] /api/answers
17:45:21:77
2023-01-08T16:45:22.267Z	undefined	ERROR	ReferenceError: exports is not defined in ES module scope
This file is being treated as an ES module because it has a '.js' file extension and '/var/task/frontend/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
    at file:///var/task/frontend/api/answers/index.js:2:23
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at async Promise.all (index 0)
    at ESMLoader.import (node:internal/modules/esm/loader:530:24)
    at importModuleDynamicallyWrapper (node:internal/vm/module:438:15)
    at getListener (/var/task/___vc/__launcher.js:37:11)
RequestId: 1251019d-68fe-469d-9b70-5c799a343823 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

Probably introduced by vercel/vercel#7954. Is there a way to downgrade the CLI?

**Comment by danielkraus** [comment]
The same. Reproduced the same error on my localhost after upgrading the Vercel CLI to [28.10.3](https://github.com/vercel/vercel/releases/tag/vercel%4028.10.3). It must be it.

**Reply by steelballs** [reply]
Yeah I did a tonne of investigation and testing and it does seem like 28.10.3 is the cause of the issue.

**Comment by styfle** [comment]
If you have tsconfig.json with `module: esnext`, then your package.json should have `type: module`.

If you don’t want to use ESM, you can change tsconfig.json to `module: commonjs`.

**Reply by duiker101** [reply]
Just tried to set `"module": "commonjs"` in tsconfig, I do not have `type: module` in my package.json. 

Still same error:
```
2023-01-08T18:59:50.401Z	undefined	ERROR	(node:8) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
2023-01-08T18:59:50.403Z	undefined	ERROR	/var/task/api/______.js:1
import { ______ } from "./______";
^^^^^^
SyntaxError: Cannot use import statement outside a module
```

**Reply by danielkraus** [reply]
The error occurs while "module: esnext" and "type: module".

**Reply by steelballs** [reply]
We tried a lot of the suggestions we found on Google, but not a single combinations worked. Would be good to hear from the Vercel team regards what changed between the CLI versions and how the deployment issues can be fixed.

**Comment by krystof-k** [comment]
I just got a response from Vercel support. It is possible to downgrade the CLI.

Set `VERCEL_CLI_VERSION` environment variable to the previous CLI verison `vercel@28.10.2`.

Builds are passing fine again.

**Reply by Lmmmmmm-bb** [reply]
Thanks!
Is this issue a bug or a feature, or is there any other information about this issue?

**Reply by krystof-k** [reply]
I've been told _"while our engineers work on an update"_ so I assume it may be a bug but we have to wait. I still think it can be because of a wrong TS configuration.

**Reply by danielkraus** [reply]
Thank you!!

**Reply by swarnava** [reply]
Please do not forget to remove the Environment variable once we fix the issue. Thank you for your patience! :)

**Reply by Palmaswell** [reply]
Thanks so much for providing a solution. I encountered a similar problem fixed by downgrading the Vercel CLI package.

**Reply by natapg** [reply]
thanks! I've been all morning trying to fix this problem

**Comment by swarnava** [comment, marked as answer]
We are **locking** this thread now to avoid confusion. As a next step:


We have an update from our engineer that may fix your issue and allow you to remove the CLI variable:
 
A `tsconfig.json` with `"module": "esnext"` (or `"module": "nodenext"`) must also have `package.json` with `"type": "module"`.

If you run into a problem, it will likely be at runtime and look something like:

```
ReferenceError: require is not defined in ES module scope, you can use import instead
```

If you want to continue using CommonJS instead of ES Modules, the solution would be to update your `tsconfig.json` file with `"module": "commonjs"`.

Also note that TypeScript will allow you to mix CJS and ESM, however Node.js will not. So you cannot use `require()` and `import()` in the same file, and you cannot use `export default` and `module.exports` in the same file.
 
If you are still facing issue with above configuration, please reach out to Vercel Support with following information:

- contents of `package.json`
- contents `tsconfig.json`
- the runtime error message (we can find this if you provide deployment url)
- the contents of the file that failed (the file will be mentioned in the error message)

If you have set any **Environment variable** to use previous version of Vercel CLI, please ensure you remove it to get future updates of our Vercel builder. Thank you!

**Reply by smaeda-ks** [reply]
Also, when switching to ESM with the above suggestion and you still see errors such as:

> Error: Cannot find module 'xxxx' imported from yyyy

Please make sure you're following the right rules...
https://www.typescriptlang.org/docs/handbook/esm-node.html

We've been seeing users are not aware of the file extension thing, particularly:
> - relative import paths need full extensions (e.g we have to write `import "./foo.js"` instead of `import "./foo"`)
